File: app/src/main/java/de/rochefort/childmonitor/VolumeHistory.java
package de.rochefort.childmonitor;

import android.os.Handler;
import android.os.Looper;
import android.support.v4.util.CircularArray;

public class VolumeHistory {
    private double mMaxVolume = 0.25;

    private double mVolumeNorm = 1.0 / mMaxVolume;
    private final CircularArray<Double> mHistory;
    private final int mMaxHistory;

    private final Handler uiHandler;

    VolumeHistory(int maxHistory) {
        uiHandler = new Handler(Looper.getMainLooper());
        mMaxHistory = maxHistory;
        mHistory = new CircularArray<>(maxHistory);
    }


    public double getVolumeNorm() {
        return mVolumeNorm;
    }

    public double get(int i) {
        return mHistory.get(i);
    }

    public int size() {
        return mHistory.size();
    }

    private void addLast(double volume) {
        // schedule editing of member vars on the ui event loop to avoid concurrency problems
        uiHandler.post(() -> {
            if (volume > mMaxVolume) {
                mMaxVolume = volume;
                mVolumeNorm = 1.0 / volume;
            }
            mHistory.addLast(volume);
            mHistory.removeFromStart(mHistory.size() - mMaxHistory);
        });
    }

    public void onAudioData(short[] data) {
        if (data.length < 1) {
            return;
        }

        final double scale = 1.0 / 128.0;
        double sum = 0;
        for (final short datum : data) {
            final double rel = datum * scale;
            sum += rel * rel;
        }
        final double volume = sum / data.length;
        addLast(volume);
    }
}

File: app/src/main/kotlin/de/rochefort/childmonitor/VolumeHistory.kt
/*
 * This file is part of Child Monitor.
 *
 * Child Monitor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Child Monitor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Child Monitor. If not, see <http://www.gnu.org/licenses/>.
 */
package de.rochefort.childmonitor

import android.os.Handler
import android.os.Looper
import androidx.collection.CircularArray

class VolumeHistory internal constructor(private val maxHistory: Int) {
    private var maxVolume = 0.25
    var volumeNorm = 1.0 / this.maxVolume
        private set
    private val historyData: CircularArray<Double> = CircularArray(maxHistory)
    private val uiHandler: Handler = Handler(Looper.getMainLooper())

    operator fun get(i: Int): Double {
        return historyData[i]
    }

    fun size(): Int {
        return historyData.size()
    }

    private fun addLast(volume: Double) {
        // schedule editing of member vars on the ui event loop to avoid concurrency problems
        uiHandler.post {
            if (volume > this.maxVolume) {
                this.maxVolume = volume
                this.volumeNorm = 1.0 / volume
            }
            historyData.addLast(volume)
            historyData.removeFromStart(historyData.size() - maxHistory)
        }
    }

    fun onAudioData(data: ShortArray) {
        if (data.isEmpty()) {
            return
        }
        val scale = 1.0 / 128.0
        var sum = 0.0
        for (datum in data) {
            val rel = datum * scale
            sum += rel * rel
        }
        val volume = sum / data.size
        addLast(volume)
    }
}
